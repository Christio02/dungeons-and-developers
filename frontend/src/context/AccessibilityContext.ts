import { createContext, useContext } from 'react';

export const AccessibilityContext = createContext({
  isAccessibilityMode: false,
  toggleAccessibilityMode: () => {},
});
export const useAccessibilityContext = () => {
  const context = useContext(AccessibilityContext);
  if (!context) {
    throw new Error('useCharacterContext must be used within a CharacterProvider');
  }
  return context;
};
