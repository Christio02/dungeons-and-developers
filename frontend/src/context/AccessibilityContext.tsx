import React, { createContext, useContext, useEffect, useState } from 'react';

const AccessibilityContext = createContext({
  isAccessibilityMode: false,
  toggleAccessibilityMode: () => {},
});

export const AccessibilityProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [isAccessibilityMode, setIsAccessibilityMode] = useState(false);
  const [isInitialized, setIsInitialized] = useState(false);

  useEffect(() => {
    const storedMode = localStorage.getItem('accesibility-mode') === 'true';
    setIsAccessibilityMode(storedMode);
    setIsInitialized(true);
  }, []);

  const toggleAccessibilityMode = () => {
    setIsAccessibilityMode((prev) => !prev);
  };

  useEffect(() => {
    if (isInitialized) {
      if (isAccessibilityMode) {
        localStorage.setItem('accesibility-mode', 'true');
        document.body.classList.add('accessibility-mode');
      } else {
        localStorage.setItem('accesibility-mode', 'false');
        document.body.classList.remove('accessibility-mode');
      }
    }
  }, [isAccessibilityMode, isInitialized]);

  return (
    <AccessibilityContext.Provider value={{ isAccessibilityMode, toggleAccessibilityMode }}>
      {children}
    </AccessibilityContext.Provider>
  );
};

export const useAccessibility = () => useContext(AccessibilityContext);
