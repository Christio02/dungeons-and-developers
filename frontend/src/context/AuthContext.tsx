import { createContext, ReactNode, useReducer } from 'react';
import { LoginState, LoginAction, LogoutAction } from '../interfaces/AuthContextProps.ts';

/**
 * AuthContext
 *
 * Provides authentication context to manage user login, logout, and related state.
 * - Stores user information such as `token`, `userId`, and `userName`.
 * - Persists authentication data in `localStorage`.
 * - Allows components to access and update authentication state via context.
 *
 * Features:
 * - Implements a `useReducer` for managing authentication state transitions.
 * - Includes `LOGIN` and `LOGOUT` actions to update state and storage.
 * - Provides helper functions for login and logout logic.
 *
 * Components:
 * - `AuthProvider`: Wraps application components to provide authentication context.
 * - `authReducer`: Reducer function for managing state transitions.
 * - `AuthContext`: Context object for accessing authentication data and methods.
 *
 * Usage:
 * - Wrap the application root with `<AuthProvider>` to enable authentication management.
 * - Access context using `useContext(AuthContext)` in child components.
 */

const initialState: LoginState = {
  token: localStorage.getItem('token') || '',
  userId: localStorage.getItem('userId') || '',
  userName: localStorage.getItem('userName') || '',
};

type AuthAction = LoginAction | LogoutAction;

const AuthContext = createContext<{
  token: string;
  userId: string;
  userName: string;
  login: (data: { token: string; userId: string; userName: string }) => void;
  logout: () => void;
}>({
  token: '',
  userId: '',
  userName: '',
  login: () => {},
  logout: () => {},
});

function authReducer(state: LoginState, action: AuthAction): LoginState {
  switch (action.type) {
    case 'LOGIN':
      return {
        token: action.payload.token,
        userId: action.payload.userId,
        userName: action.payload.userName,
      };
    case 'LOGOUT':
      return {
        token: '',
        userId: '',
        userName: '',
      };
    default:
      return state;
  }
}

function AuthProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(authReducer, initialState);

  const login = (userData: { token: string; userId: string; userName: string }) => {
    console.log('Logging in with:', userData);
    localStorage.setItem('token', userData.token);
    localStorage.setItem('userId', userData.userId);
    localStorage.setItem('userName', userData.userName);
    localStorage.setItem('loginToast', 'true');
    dispatch({ type: 'LOGIN', payload: userData });
  };

  const logout = () => {
    localStorage.clear();
    dispatch({ type: 'LOGOUT' });
  };

  return (
    <AuthContext.Provider value={{ token: state.token, userId: state.userId, userName: state.userName, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
}

export { AuthContext, AuthProvider };
