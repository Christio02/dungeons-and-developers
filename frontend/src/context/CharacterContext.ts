import { createContext, useContext } from 'react';
import { CharacterContextType } from '../interfaces/CharacterContextProps.ts';

export const CharacterContext = createContext<CharacterContextType>({
  stateAbilities: [],
  userAbilityScores: new Map<string, number>(),
  updateAbilityScores: async () => Promise.resolve(),
  classes: [],
  updateClass: async () => Promise.resolve(),
  races: [],
  updateRace: async () => Promise.resolve(),
  addToEquipments: () => {},
  removeFromEquipments: () => {},
  removeAllEquipments: () => {},
  loadingStates: {
    equipmentsLoading: false,
    classLoading: false,
    abilityScoresLoading: false,
    raceLoading: false,
  },
});

export const useCharacterContext = () => {
  const context = useContext(CharacterContext);
  if (!context) {
    throw new Error('useCharacterContext must be used within a CharacterProvider');
  }
  return context;
};
