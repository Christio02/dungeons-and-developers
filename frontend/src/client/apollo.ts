import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import { ReviewType } from '../interfaces/ReviewProps.ts';

/**
 * Apollo Client for interacting with GraphQL server
 * Implements authorization, caching
 * **/
const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});
/**
 * Configures authorization by given the 'Bearer' token to the request header.
 **/
const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

/**
 * Apollo Client configuration.
 * With custom cache behaviour for types and fields.
 * Implements custom Merge behaviours
 **/
const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
          equipments: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Ability: {
        keyFields: ['id'],
      },
      Equipment: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
        fields: {
          reviews: {
            merge(existing = [], incoming) {
              const mergedReviews = [...existing];
              incoming.forEach((incomingReview: ReviewType) => {
                if (!mergedReviews.some((review) => review.id === incomingReview.id)) {
                  mergedReviews.push(incomingReview);
                }
              });
              return mergedReviews;
            },
          },
        },
      },
      Review: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
