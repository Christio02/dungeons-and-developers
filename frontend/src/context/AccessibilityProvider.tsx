import { useEffect, useState } from 'react';
import { AccessibilityContext } from './AccessibilityContext';
import { AccessibilityProviderProps } from '../interfaces/AccessibilityProviderProps.ts';

export const AccessibilityProvider = ({ children }: AccessibilityProviderProps) => {
  const [isAccessibilityMode, setIsAccessibilityMode] = useState(false);
  const [isInitialized, setIsInitialized] = useState(false);

  useEffect(() => {
    const storedMode = localStorage.getItem('accessibility-mode') === 'true';
    setIsAccessibilityMode(storedMode);
    setIsInitialized(true);
  }, []);

  const toggleAccessibilityMode = () => {
    setIsAccessibilityMode((prev) => !prev);
  };

  useEffect(() => {
    if (isInitialized) {
      if (isAccessibilityMode) {
        localStorage.setItem('accessibility-mode', 'true');
        document.body.classList.add('accessibility-mode');
      } else {
        localStorage.setItem('accessibility-mode', 'false');
        document.body.classList.remove('accessibility-mode');
      }
    }
  }, [isAccessibilityMode, isInitialized]);

  return (
    <AccessibilityContext.Provider value={{ isAccessibilityMode, toggleAccessibilityMode }}>
      {children}
    </AccessibilityContext.Provider>
  );
};
