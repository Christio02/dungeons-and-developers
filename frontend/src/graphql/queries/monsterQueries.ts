import { gql } from '@apollo/client';

/**
 * Monster Queries
 *
 * Features:
 * - Fetches paginated and filtered lists of monsters.
 * - Retrieves monster type counts and hit point ranges for filters.
 * - Provides monster suggestions for search and autocomplete.
 * - Fetches reviews associated with a specific monster.
 */

/**
 * - `GET_MONSTERS`: Fetches a paginated list of monsters based on various filter options.
 *   - Query Parameters:
 *     - `searchTerm` (optional): Filter monsters by name or related terms.
 *     - `offset` (optional): Number of records to skip for pagination.
 *     - `limit` (optional): Maximum number of records to retrieve.
 *     - `types` (optional): Filter by a list of monster types.
 *     - `minHp` (optional): Minimum hit points for filtering.
 *     - `maxHp` (optional): Maximum hit points for filtering.
 *     - `sortOption` (optional): Sorting option for monsters.
 *   - Query Response:
 *     - `monsters`: List of monster objects with:
 *       - `id`: Unique identifier.
 *       - `name`: Name of the monster.
 *       - `size`: Monster size category.
 *       - `type`: Type of monster.
 *       - `alignment`: Alignment of the monster.
 *       - `hit_points`: Total hit points.
 *       - `image`: URL of the monster's image.
 *       - `averageDifficulty`: Average difficulty rating.
 *       - `reviewsCount`: Number of reviews.
 *     - `totalMonsters`: Total number of monsters available.
 *     - `minHp` and `maxHp`: Minimum and maximum hit points across all monsters.
 */

export const GET_MONSTERS = gql`
  query GetMonsters(
    $searchTerm: String
    $offset: Int
    $limit: Int
    $types: [String!]
    $minHp: Int
    $maxHp: Int
    $sortOption: String
  ) {
    monsters(
      searchTerm: $searchTerm
      offset: $offset
      limit: $limit
      types: $types
      minHp: $minHp
      maxHp: $maxHp
      sortOption: $sortOption
    ) {
      monsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
        averageDifficulty
        reviewsCount
      }
      totalMonsters
      minHp
      maxHp
    }
  }
`;

/**
 * - `GET_MONSTER_TYPE_COUNTS`: Fetches counts of monster types for given hit point ranges.
 *   - Query Parameters:
 *     - `minHp` (optional): Minimum hit points for filtering.
 *     - `maxHp` (optional): Maximum hit points for filtering.
 *   - Query Response:
 *     - `monsterTypeCounts`: Array of types with:
 *       - `type`: Monster type.
 *       - `count`: Count of monsters of this type.
 */
export const GET_MONSTER_TYPE_COUNTS = gql`
  query GetMonsterTypeCounts($minHp: Int, $maxHp: Int) {
    monsterTypeCounts(minHp: $minHp, maxHp: $maxHp) {
      type
      count
    }
  }
`;

/**
 * - `GET_MONSTER_HP_RANGE`: Retrieves the global range of hit points for all monsters.
 *   - Query Response:
 *     - `monsterHpRange`: Object with:
 *       - `minHp`: Minimum hit points.
 *       - `maxHp`: Maximum hit points.
 */
export const GET_MONSTER_HP_RANGE = gql`
  query GetMonsterHpRange {
    monsterHpRange {
      minHp
      maxHp
    }
  }
`;

/**
 * - `GET_MONSTER_SUGGESTIONS`: Provides monster suggestions for autocomplete or filtering.
 *   - Query Parameters:
 *     - `searchTerm` (optional): Filter suggestions by term.
 *     - `types` (optional): Restrict suggestions to specific monster types.
 *     - `minHp` (optional): Minimum hit points for filtering.
 *     - `maxHp` (optional): Maximum hit points for filtering.
 *   - Query Response:
 *     - `monsters`: List of suggested monsters with:
 *       - `id`: Unique identifier.
 *       - `name`: Monster name.
 */
export const GET_MONSTER_SUGGESTIONS = gql`
  query GetMonsterSuggestions($searchTerm: String, $types: [String!], $minHp: Int, $maxHp: Int) {
    monsters(searchTerm: $searchTerm, types: $types, minHp: $minHp, maxHp: $maxHp) {
      monsters {
        id
        name
      }
    }
  }
`;

/**
 * - `GET_MONSTER_REVIEWS`: Fetches reviews associated with a specific monster.
 *   - Query Parameters:
 *     - `monsterId` (required): Unique identifier of the monster.
 *   - Query Response:
 *     - `monster`: Object containing:
 *       - `id`: Monster identifier.
 *       - `reviews`: List of review objects with:
 *         - `id`: Review identifier.
 *         - `user`: Reviewer details with:
 *           - `id`: User identifier.
 *           - `userName`: Name of the reviewer.
 *         - `difficulty`: Difficulty rating of the monster.
 *         - `description`: Text of the review.
 *         - `createdAt`: Timestamp of the review creation.
 */
export const GET_MONSTER_REVIEWS = gql`
  query GetMonsterReviews($monsterId: ID!) {
    monster(id: $monsterId) {
      id
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;
