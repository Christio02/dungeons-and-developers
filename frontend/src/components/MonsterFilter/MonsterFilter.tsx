import { useQuery } from '@apollo/client';
import { Box, Slider } from '@mui/material';
import { useEffect, useRef, useState } from 'react';
import { FaChevronDown } from 'react-icons/fa';
import { FiX } from 'react-icons/fi';
import { GET_MONSTER_HP_RANGE } from '../../graphql/queries/monsterQueries.ts';
import { useToast } from '../../hooks/useToast.ts';
import CustomCheckbox from '../CustomCheckbox/CustomCheckbox.tsx';

interface MonsterFilterProps {
  selectedFilters: Set<string>;
  setSelectedFilters: (filters: Set<string> | ((prev: Set<string>) => Set<string>)) => void;
  onHpChange: (min: number, max: number) => void;
  setCurrentPage: (page: number) => void;
  onClearFilters: () => void;
  monsterCounts: Record<string, number>;
  monsterTypes: string[];
  searchTerm: string;
}

export default function MonsterFilter({
  selectedFilters,
  setSelectedFilters,
  onHpChange,
  setCurrentPage,
  onClearFilters,
  monsterCounts,
  monsterTypes,
  searchTerm,
}: MonsterFilterProps) {
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const { data: hpRangeData, loading: hpRangeLoading } = useQuery(GET_MONSTER_HP_RANGE);
  const initialHpRange = hpRangeData ? [hpRangeData.monsterHpRange.minHp, hpRangeData.monsterHpRange.maxHp] : [1, 546];
  const [hpRange, setHpRange] = useState<number[]>(initialHpRange);
  const [outOfRangeFilters, setOutOfRangeFilters] = useState<Set<string>>(new Set());
  const dropdownRef = useRef<HTMLDivElement>(null);
  const toastContext = useToast();
  const showToast = toastContext ? toastContext.showToast : () => {};

  useEffect(() => {
    if (hpRangeData && !hpRangeLoading) {
      setHpRange([hpRangeData.monsterHpRange.minHp, hpRangeData.monsterHpRange.maxHp]);
    }
  }, [hpRangeData, hpRangeLoading]);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
        setIsDropdownOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [dropdownRef]);

  useEffect(() => {
    setOutOfRangeFilters((prevOutOfRange) => {
      const updatedOutOfRange = new Set(prevOutOfRange);
      let modified = false;

      selectedFilters.forEach((filter) => {
        if (monsterCounts[filter] === 0) {
          updatedOutOfRange.add(filter);
          modified = true;
        } else if (updatedOutOfRange.has(filter)) {
          updatedOutOfRange.delete(filter);
          modified = true;
        }
      });

      return modified ? updatedOutOfRange : prevOutOfRange;
    });
  }, [monsterCounts, selectedFilters]);

  const handleCheckboxChange = (option: string) => {
    setSelectedFilters((prev: Set<string>) => {
      const newFilters = new Set<string>(prev);
      const outOfRange = new Set(outOfRangeFilters);

      if (newFilters.has(option)) {
        newFilters.delete(option);
        outOfRange.delete(option);
      } else {
        newFilters.add(option);
        if (monsterCounts[option] === 0) {
          outOfRange.add(option);
        }
      }

      setOutOfRangeFilters(outOfRange);
      return newFilters;
    });
    setCurrentPage(1);
  };

  const toggleDropdown = () => {
    if (searchTerm) {
      showToast({
        message: 'Please clear the search bar before using the filters.',
        type: 'warning',
        duration: 4000,
      });
    } else {
      setIsDropdownOpen((prev) => !prev);
    }
  };

  const handleSliderChange = (_: Event, newValue: number | number[]) => {
    const [min, max] = newValue as number[];
    setHpRange([min, max]);
    onHpChange(min, max);
  };

  const handleClearFilters = () => {
    if (selectedFilters.size > 0 || hpRange[0] !== initialHpRange[0] || hpRange[1] !== initialHpRange[1]) {
      setSelectedFilters(new Set<string>());
      setHpRange(initialHpRange);
      onHpChange(initialHpRange[0], initialHpRange[1]);
      setOutOfRangeFilters(new Set());
      onClearFilters();
    }
  };

  return (
    <div className="relative text-white" ref={dropdownRef}>
      <button
        onClick={toggleDropdown}
        className={`sub-header px-1 rounded-md border-2 transition-colors duration-200 text-nowrap ${
          searchTerm
            ? 'bg-customGray text-gray-500 border-gray-500 cursor-not-allowed'
            : 'bg-customRed hover:bg-transparent border-customRed hover:border-customRed hover:text-customRed'
        }`}
        style={{ pointerEvents: searchTerm ? 'auto' : 'initial' }}
      >
        Filter Monsters
        <FaChevronDown className="inline ml-3" />
      </button>
      {isDropdownOpen && (
        <div
          className="absolute bg-customGray shadow-xl shadow-black p-8 rounded mt-2 min-w-[30vw] w-max z-50 min-h-fit
        xl:top-1/2 xl:left-2 xl:transform xl:-translate-x-2 xl:-translate-y-2
        left-1/2 transform -translate-x-1/2 space-y-6"
          style={{ maxHeight: '80vh', overflowY: 'auto' }}
        >
          <div className="flex gap-3 items-center justify-between">
            <button
              onClick={handleClearFilters}
              className="transition-colors hover:text-customRed sub-header focus:outline-none focus-visible:ring-2 focus-visible:ring-customRed"
            >
              Clear Filters
            </button>
            <FiX
              className="h-8 w-8 text-white hover:text-customRed cursor-pointer focus:outline-none focus-visible:ring-2 focus-visible:ring-customRed"
              onClick={toggleDropdown}
              tabIndex={0}
              aria-label="Close"
              onKeyDown={(e) => {
                if (e.key === 'Enter' || e.key === ' ') {
                  toggleDropdown();
                }
              }}
            />
          </div>
          <Box>
            <h2 className="text bold">Hit Points:</h2>
            <Slider
              value={hpRange}
              onChange={handleSliderChange}
              valueLabelDisplay="auto"
              min={initialHpRange[0]}
              max={initialHpRange[1]}
              sx={{
                '& .MuiSlider-thumb': { color: '#DB3232', width: 24, height: 24 },
                '& .MuiSlider-track': { color: '#DB3232', height: 10 },
                '& .MuiSlider-rail': { color: '#DB3232', height: 10 },
              }}
            />
            <Box display="flex" justifyContent="space-between" mt={1}>
              <span className="xs:text-xs sm:text-sm md:text-md lg:text-lg xl:text-xl">min: {hpRange[0]}</span>
              <span className="xs:text-xs sm:text-sm md:text-md lg:text-lg xl:text-xl">max: {hpRange[1]}</span>
            </Box>
          </Box>
          <h2 className="text bold mb-2">Type:</h2>
          <div className="grid grid-cols-1 gap-x-8 gap-8 md:grid-cols-2 xl:grid-cols-3">
            {[...new Set([...monsterTypes, ...Array.from(selectedFilters)])].map((option) => {
              const count = monsterCounts[option] || 0;
              const isDisabled = count === 0 && !selectedFilters.has(option);

              return (
                <label
                  key={option}
                  className={`flex items-center gap-x-6 text ${isDisabled ? 'text-gray-500' : ''}`}
                  style={{ cursor: isDisabled ? 'not-allowed' : 'pointer' }}
                >
                  <CustomCheckbox
                    key={`${option}-${selectedFilters.has(option)}`}
                    checked={selectedFilters.has(option)}
                    onChange={() => handleCheckboxChange(option)}
                    scale={0.8}
                    disabled={isDisabled}
                  />
                  <span className="text">
                    {option} ({count})
                  </span>
                </label>
              );
            })}
          </div>
        </div>
      )}
    </div>
  );
}
