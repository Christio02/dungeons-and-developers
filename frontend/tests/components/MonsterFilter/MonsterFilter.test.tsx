import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MonsterFilter from '../../../src/components/MonsterFilter/MonsterFilter';
import { describe, it, expect, vi, afterEach } from 'vitest';
import { MockedProvider } from '@apollo/client/testing';
import { AuthContext } from '../../../src/context/AuthContext';
import { GET_MONSTER_HP_RANGE, GET_MONSTER_TYPE_COUNTS } from '../../../src/graphql/queries/monsterQueries';

const mockAuthContextValue = { userId: '12345', userName: 'Bob', token: '32', login: vi.fn(), logout: vi.fn() };

const mockShowToast = vi.fn();
vi.mock('../../../src/hooks/useToast.ts', () => ({
  useToast: () => ({
    showToast: mockShowToast,
  }),
}));

const mockSetSelectedFilters = vi.fn();

let selectedFilters = new Set<string>();
const mockMonsterCounts = {
  dragon: 5,
  fiend: 0,
  humanoid: 8,
};

const mockMonsterTypes = ['dragon', 'fiend', 'humanoid'];

const mockOnClearFilters = () => {};
const mockOnHpChange = () => {};
const mockSetCurrentPage = () => {};

const searchTerm = '';

const mocks = [
  {
    request: {
      query: GET_MONSTER_HP_RANGE,
    },
    result: {
      data: {
        monsterHpRange: {
          minHp: 10,
          maxHp: 50,
          __typename: 'MonsterHpRange',
        },
      },
    },
  },
  {
    request: {
      query: GET_MONSTER_TYPE_COUNTS,
      variables: { minHp: 10, maxHp: 100 },
    },
    result: {
      data: {
        monsterTypeCounts: [
          { type: 'dragon', count: 5 },
          { type: 'fiend', count: 0 },
          { type: 'humanoid', count: 8 },
        ],
      },
    },
  },
];
describe('MonsterFilter Component', () => {
  afterEach(() => {
    vi.restoreAllMocks();
    selectedFilters = new Set();
  });

  it('renders correctly and matches snapshot', () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider value={mockAuthContextValue}>
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
            searchTerm={searchTerm}
            monsterTypes={mockMonsterTypes}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('opens and closes the dropdown menu', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider value={mockAuthContextValue}>
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
            searchTerm={searchTerm}
            monsterTypes={mockMonsterTypes}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );
    const filterButton = screen.getByText('Filter Monsters');

    await userEvent.click(filterButton);

    await waitFor(() => {
      expect(screen.getByText('Clear Filters')).toBeTruthy();
    });

    const closeButton = screen.getByLabelText('Close');
    await userEvent.click(closeButton);
    await waitFor(() => {
      expect(screen.queryByText('Clear Filters')).not.toBeTruthy();
    });
  });

  it('fetches and displays monster types correctly', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider value={mockAuthContextValue}>
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
            searchTerm={searchTerm}
            monsterTypes={mockMonsterTypes}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );

    const filterButton = screen.getByText('Filter Monsters');
    await userEvent.click(filterButton);

    await waitFor(() => {
      expect(screen.getByText('dragon (5)')).toBeTruthy();
      expect(screen.getByText('fiend')).toBeTruthy();
      expect(screen.getByText('humanoid (8)')).toBeTruthy();
    });
  });

  it('toggles filters and updates selected filters correctly', async () => {
    render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider value={mockAuthContextValue}>
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
            searchTerm={searchTerm}
            monsterTypes={mockMonsterTypes}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );
    const filterButton = screen.getByText('Filter Monsters');

    await userEvent.click(filterButton);
    await waitFor(() => {
      expect(screen.getByText('dragon (5)')).toBeTruthy();
    });

    const dragonLabel = screen.getByText('dragon (5)');
    await userEvent.click(dragonLabel);
    await waitFor(() => {
      expect(mockSetSelectedFilters).toHaveBeenCalledTimes(1);
      const updaterFunction = mockSetSelectedFilters.mock.calls[0][0];
      const updatedSet = updaterFunction(new Set());
      expect(updatedSet).toContain('dragon');
    });

    await userEvent.click(dragonLabel);
    await waitFor(() => {
      expect(mockSetSelectedFilters).toHaveBeenCalledTimes(2);
      const updaterFunction = mockSetSelectedFilters.mock.calls[1][0];
      const updatedSet = updaterFunction(new Set(['dragon']));
      expect(updatedSet).not.toContain('dragon');
    });
  });

  it('clears all selected filters', async () => {
    selectedFilters = new Set(['dragon', 'humanoid']);
    render(
      <MockedProvider mocks={mocks}>
        <AuthContext.Provider value={mockAuthContextValue}>
          <MonsterFilter
            selectedFilters={selectedFilters}
            setSelectedFilters={mockSetSelectedFilters}
            monsterCounts={mockMonsterCounts}
            onClearFilters={mockOnClearFilters}
            onHpChange={mockOnHpChange}
            setCurrentPage={mockSetCurrentPage}
            searchTerm={searchTerm}
            monsterTypes={[]}
          />
        </AuthContext.Provider>
      </MockedProvider>
    );
    const filterButton = screen.getByText('Filter Monsters');
    await userEvent.click(filterButton);

    const clearButton = screen.getByText('Clear Filters');
    await userEvent.click(clearButton);

    expect(mockSetSelectedFilters).toHaveBeenCalledTimes(1);
    expect(mockSetSelectedFilters).toHaveBeenCalledWith(new Set());
  });
});
