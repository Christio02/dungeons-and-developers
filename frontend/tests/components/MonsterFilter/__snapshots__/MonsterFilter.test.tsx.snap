// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MonsterFilter Component > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="relative text-white"
  >
    <button
      class="sub-header px-1 rounded-md border-2 transition-colors duration-200 text-nowrap bg-customRed hover:bg-transparent border-customRed hover:border-customRed hover:text-customRed"
      style="pointer-events: initial;"
    >
      Filter Monsters
      <svg
        class="inline ml-3"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </button>
  </div>
</div>
`;
